be/src/storage/lake/tablet_manager.cpp

```diff
@@ -627,27 +627,29 @@ void TabletManager::update_metacache_limit(size_t new_capacity) {
 }
 
 int64_t TabletManager::in_writing_data_size(int64_t tablet_id) {
-    int64_t size = 0;
-    std::shared_lock rdlock(_meta_lock);
-    const auto& it = _tablet_in_writing_size.find(tablet_id);
-    if (it != _tablet_in_writing_size.end()) {
-        size = it->second;
+    {
+        std::shared_lock rdlock(_meta_lock);
+        const auto& it = _tablet_in_writing_size.find(tablet_id);
+        if (it != _tablet_in_writing_size.end()) {
+            VLOG(1) << "tablet " << tablet_id << " in writing data size: " << it->second;
+            return it->second;
+        }
     }
-    VLOG(1) << "tablet " << tablet_id << " in writing data size: " << size;
-    return size;
+    return add_in_writing_data_size(tablet_id, 0);
 }
 
-void TabletManager::add_in_writing_data_size(int64_t tablet_id, int64_t size) {
+int64_t TabletManager::add_in_writing_data_size(int64_t tablet_id, int64_t size) {
     std::unique_lock wrlock(_meta_lock);
-    _tablet_in_writing_size[tablet_id] += size;
+    const auto& it = _tablet_in_writing_size.find(tablet_id);
+    if (it != _tablet_in_writing_size.end()) {
+        it->second += size;
+    } else {
+        // data size is not in cache, get it from tablet meta
+        _tablet_in_writing_size[tablet_id] = get_tablet_data_size(tablet_id, nullptr).value_or(0) + size;
```

Comment:
`get_tablet_data_size` will possibly trigger a remote IO. it will hold the write lock for a long time, shall release the lock and get the size back and redo the locking to update the size.